Try to add BFLanguage and instrumentation support

tmysik opened the pull request:
It would be great to get any feedback (no need to hurry, of course).

Thanks!

_Note:_ Inspired from SL, of course. Problematic parts marked with ```XXX```.

tmysik committed:
Add BFLanguage & use it

However, output does not work, not sure why (likely some my stupid mistake).

tmysik committed:
Add instrumentation support

Unfortunately, debugger does not work, NPE occurs.

chumer commented on src/main/java/com/oracle/truffle/bf/test/OperationNode.java:
```diff
@@ -0,0 +1,119 @@
+        }
+        if (tag == DebuggerTags.AlwaysHalt.class) {
+            return true;
+        }
```
I don't think you want to always halt on each statement. This is for language constructs that imply a breakpoint.

tmysik replied:
I know. I just wanted to test debugger, to be stopped at the very first statement (not sure if it would work).

chumer commented on src/main/java/com/oracle/truffle/bf/test/BFRootNode.java:
```diff
@@ -0,0 +1,58 @@
+import com.oracle.truffle.api.nodes.RootNode;
+import com.oracle.truffle.api.source.SourceSection;
+
+@Instrumentable(factory = BFRootNodeWrapper.class)
```
RootNode classes cannot be instrumented, because they cannot be replaced. Should actually produce an error, as far as I implemented that. If not we have found an issue.

chumer commented on src/main/java/com/oracle/truffle/bf/test/BFMain.java:
```diff
@@ -0,0 +1,150 @@
+public final class BFMain {
+
+    // XXX debug does not work, NPE
+    private static final boolean DEBUG = false;
```
You could try to write a DebuggerTest using the Truffle TCK DebuggerTester class.
See for example:
https://github.com/graalvm/truffle/blob/master/truffle/com.oracle.truffle.sl.test/src/com/oracle/truffle/sl/test/SLDebugTest.java

tmysik replied:
Thanks, will look at it!

chumer commented on src/main/java/com/oracle/truffle/bf/test/BFLanguage.java:
```diff
@@ -0,0 +1,87 @@
+    protected CallTarget parse(Source source, Node node, String... strings) throws Exception {
+        BFParser.Operation[] operations = new BFParser().parse(source.getInputStream());
+        BFRootNode rootNode = new BFRootNode(prepareNodes(source, operations), source.createSection(0, source.getLength()));
+        return Truffle.getRuntime().createCallTarget(rootNode);
```
What you also want for debugging, you want SourceSections for each Node by implementing Node#getSourceSection().

tmysik replied:
There already is implementation of `getSourceSection()` on `OperationNode` - am I missing something?

chumer commented on src/main/java/com/oracle/truffle/bf/test/BFLanguage.java:
```diff
@@ -0,0 +1,87 @@
+    @Override
+    protected CallTarget parse(Source source, Node node, String... strings) throws Exception {
+        BFParser.Operation[] operations = new BFParser().parse(source.getInputStream());
+        BFRootNode rootNode = new BFRootNode(prepareNodes(source, operations), source.createSection(0, source.getLength()));
```
You could try to merge the BFParser with the prepareNodes step. Would make it easier to assign source sections.

tmysik replied:
You mean to be able to get `OperationNode` instances directly from `BFParser`? Thanks!

chumer commented on src/main/java/com/oracle/truffle/bf/test/BFContext.java:
```diff
@@ -0,0 +1,34 @@
+ */
+package com.oracle.truffle.bf.test;
+
+import com.oracle.truffle.api.ExecutionContext;
```
You could try to do some interop with JavaScript by using Interop.eval(mimeType, script).

tmysik replied:
Any hint? I am afraid I do not fully understand what you mean exactly, sorry. Thanks!

tmysik replied:
Or an example what you mean.

tmysik committed:
Remove instrumentation from root node

RootNode classes cannot be instrumented, because they cannot be replaced.

tmysik committed:
Copy BFParser to the test package

To be able to easily modify it without changing the original parser, if needed.
